### PR TITLE
Some refactoring so we arent checking undefined on program or asserting it so much

### PR DESCRIPTION
**Pull request by sheetalkamat**

### CODE REVIEW

File: src/compiler/watch.ts [hunk 1]
```diff
@@ -329,8 +329,8 @@ function createTabularErrorsDisplay(filesInError: (ReportFileInError | undefined
 }
 
 /** @internal */
-export function isBuilderProgram(program: Program | BuilderProgram): program is BuilderProgram {
-    return !!(program as BuilderProgram).getState;
+export function isBuilderProgram<T extends BuilderProgram>(program: Program | BuilderProgram): program is T {
```
Comment (jakebailey) [inline src/compiler/watch.ts:332]: This seems unsafe; this could conform to anything, right?

Comment (jakebailey) [reply src/compiler/watch.ts:332]: I guess it's copied from somewhere else...

**Review by jakebailey — APPROVED** [review]
Reminds me that we should see if we can make `Debug.checkDefined` or `Debug.assert` reject always-defined or never-falsy values (statically or at lint time), in case there are more of these that could be removed.

File: src/compiler/watch.ts (hunk 1 above)
Comment (sheetalkamat) [reply src/compiler/watch.ts:332]: Yep

**Pull request merged**